joblib/parallel.py

```diff
@@ -1276,31 +1273,30 @@ def _get_outputs(self, iterator, pre_dispatch):
             # the user if necessary.
             self._exception = True
 
-            # In CPython, the gc and the error happen in the main thread.
-            # However, this is not the case for other interpreters such as in
-            # PyPy, where this block can be executed in arbitrary threads. This
-            # can lead to hang when a process tries to join itself. As a
+            # In some interpreters such as PyPy or potentially CPython without
+            # GIL, this error might happen not in the main thread. This
+            # can lead to hang when a thread tries to join itself. As a
             # workaround, we detach the execution of the aborting code to a
             # dedicated thread. We then need to make sure the rest of the
             # function does not call `_terminate_and_reset` in finally.
```

Comment:
It does not really makes sense to reason in terms of "main" thread. A joblib parallel call can be started from a non-main thread.

```suggestion
            # In some interpreters such as PyPy, GeneratorExit can be raised in
            # a different thread than the one used to start the dispatch of the
            # parallel tasks. This can lead to hang when a thread attempts to
            # join itself. As workaround, we detach the execution of the aborting
            # code to a dedicated thread. We then need to make sure the rest of the
            # function does not call `_terminate_and_reset` in finally.
```